### PR TITLE
Add support for go-getter modes

### LINKED ISSUE #2678: Control artifact go-getter "mode"
### Nomad version

Nomad v0.5.6

### Issue

When using go-getter within s3, to download a specific file, deeply nested inside a directory hierarchy, the current usage of go-getter requires some unnecessary wide s3 policies to work.

With the current usage, you need a policy similar to the one below to make downloading artficats work, since go-getter run in [`any Mode`](https://github.com/hashicorp/nomad/blob/master/client/getter/getter.go#L48) by default - meaning it will try to traverse the bucket to find the file, and other files (if it happen to be a directory)

Being able to control the `Mode` would allow us to drop the `ListObjects` policy from our aws instances, and further harden our security footprint. 

Looking at the code in go-getter, it seem like it tries, by default, to be recursive and download multiple files in the bucket if the `source` in nomad artifact happened to be a folder rather than a file

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::bucket/folder1/folder2/folder3/file"
            ],
            "Effect": "Allow"
        },
        {
            "Action": [
                "s3:ListObjects"
            ],
            "Resource": [
                "arn:aws:s3:::bucket/*",
                "arn:aws:s3:::bucket"
            ],
            "Effect": "Allow"
        }
    ]
}
```

#### any mode (current in nomad)

Without `ListObjects` policy:

```
go-getter -mode any s3::https://s3.amazonaws.com/bucket/folder1/folder2/folder3/file file
2017/05/30 10:16:06 Error downloading: AccessDenied: Access Denied
```

WIth `ListObjects` policy:

```
go-getter -mode any s3::https://s3.amazonaws.com/bucket/folder1/folder2/folder3/file file
2017/05/30 10:16:06 OK!
```

#### dir mode 

Without `ListObjects` policy:

```
go-getter -mode dir s3::https://s3.amazonaws.com/bucket/folder1/folder2/folder3/file file
2017/05/30 10:16:06 Error downloading: AccessDenied: Access Denied
```

WIth `ListObjects` policy:

```
go-getter -mode dir s3::https://s3.amazonaws.com/bucket/folder1/folder2/folder3/file file
2017/05/30 10:16:06 OK!
```

#### file mode 

Without `ListObjects` policy:

```
go-getter -mode file s3::https://s3.amazonaws.com/bucket/folder1/folder2/folder3/file file
2017/05/30 10:16:06 OK
```

With `ListObjects` policy:

```
go-getter -mode file s3::https://s3.amazonaws.com/bucket/folder1/folder2/folder3/file file
2017/05/30 10:16:06 OK
```

### PR DESCRIPTION
**Pull request by schmichael**
Fixes #2678 

Also updates go-getter to hashicorp/go-getter#69 so we now support S3 alternatives.

### CODE REVIEW

**Comment by lfarnell** [comment]
@schmichael I would note for the minio support, you must explicitly set the protocol to S3. I can send a separate pr to update the docs if you want or you can just include it here. I should also update the readme for go-getter

**Comment by schmichael** [comment]
@lfarnell Good call! I'm updating Nomad's docs right now. Mind doing the PR against go-getter's README?

Thanks!

**Comment by lfarnell** [comment]
@schmichael So I must have updated the README for go-getter and not remembered. All good to go :smile:

**Comment by jippi** [comment]
Thanks for this PR @schmichael - will make stuff a lot simpler for us going forward! 0.6 will be an epic release by the look of it

File: CHANGELOG.md [hunk 1]
```diff
@@ -44,6 +44,8 @@ BUG FIXES:
    allocation destroy fixing a corner case in which an allocation may be blocked
    till destroy [GH-2563]
  * client/artifact: Honor netrc [GH-2524]
+ * client/artifact: Allow specifying a go-gettter mode [GH-2781]
```
Comment (dadgar) [inline CHANGELOG.md]: spelttt wrong

**Pull request merged**